### ISSUE TITLE
callTracer reports inaccurate gasUsed

### ISSUE THREAD

**Issue by geomad** [ethereum/go-ethereum#27746]
#### System information
Geth version:  v 1.12.0

The gas reported from `debug_traceCall` with `callTracer` is often innacurate.
I wrote a script that takes the `index 0` tx of every block and then call `callTracer` on it on the previous block, which should almost always result in the same gas and here are the results:
```
actual gas: 201547 trace gas: 114291
actual gas: 33952 trace gas: 33952
actual gas: 160332 trace gas: 149916
actual gas: 350319 trace gas: 170771
actual gas: 174309 trace gas: 174309
actual gas: 184034 trace gas: 184034
actual gas: 255780 trace gas: 247448
actual gas: 21000 trace gas: 21000
actual gas: 213557 trace gas: 121879
```
The script: 
```
web3.eth.subscribe('newBlockHeaders')
    .on("data", async function (blockHeader) {
      let block = await web3.eth.getBlock(blockHeader.number);
      let firstTx = await web3.eth.getTransaction(block.transactions[0])
      let firstTxReceipt = await web3.eth.getTransactionReceipt(firstTx.hash)
      let tx = {
        from: firstTx.from,
        to: firstTx.to,
        value: web3.utils.toHex(firstTx.value),
        data: firstTx.input,
        type: web3.utils.toHex(firstTx.type)
      }
      if (firstTx.accessList) tx.accessList = firstTx.accessList
      let trace = await web3.debug.traceCall(tx, web3.utils.toHex(blockHeader.number - 1), {
        tracer: 'callTracer',
        tracerConfig: { onlyTopCall: true }
      })
      console.log("actual gas:", firstTxReceipt.gasUsed.toString(), "trace gas:", new BN(trace.gasUsed).toString())
    })
```

**Comment by s1na** [comment]
My guess is it is inaccurate because you're not providing the `gas` field. Another way to test this is to get the txhash for one of those failing txes and do `debug.traceTransaction` if that reports an inaccurate gas too that indicates a bug.

**Comment by holiman** [comment]
IMO it's a bit handwavy to just say "should almost always result in the same gas", and take the fact that it appears not to, and attribute that to a bug in geth. 
Please find a demo where you can demonstrate somehow that what happens is indeed an error. For example a particular transaction hash which reports inaccurate results this way.

**Comment by geomad** [comment]
@s1na @holiman thank you!
here is an example with a specific transaction hash, `debug.traceTransaction` and `gas` field added.

Result:
`actual gas: 181192 callTracer gas: 170776 traceTransaction gas 181192`
Script:
```
  const transaction = '0xead3428a741f5d012d7203b086b2ffb32e1f1c3536779d62ae129cbffb0908c3'
  let firstTx = await web3.eth.getTransaction(transaction)
  let firstTxReceipt = await web3.eth.getTransactionReceipt(firstTx.hash)
  let tx = {
    from: firstTx.from,
    to: firstTx.to,
    gas: web3.utils.toHex(firstTx.gas),
    value: web3.utils.toHex(firstTx.value),
    data: firstTx.input,
    type: web3.utils.toHex(firstTx.type)
  }
  if (firstTx.accessList) tx.accessList = firstTx.accessList
  let trace = await web3.debug.traceCall(tx, web3.utils.toHex(firstTx.blockNumber - 1), {
    tracer: 'callTracer',
    tracerConfig: { onlyTopCall: true }
  })
  let traceTransaction = await web3.debug.traceTransaction(firstTx.hash)
  console.log("actual gas:", firstTxReceipt.gasUsed.toString(), "callTracer gas:", new BN(trace.gasUsed).toString(), "traceTransaction gas", traceTransaction.gas.toString())

```

**Comment by jsvisa** [comment]
I can reproduce this issue, and here is a simple example to reproduce this issue:


```python
#!/usr/bin/env python
import requests

# tx: 0xead3428a741f5d012d7203b086b2ffb32e1f1c3536779d62ae129cbffb0908c3
req = {
    "jsonrpc": "2.0",
    "method": "debug_traceCall",
    "id": 67,
    "params": [
        {
            "from": "0x0C46dADBc6271511F25273cE856eCF799022dA63",
            "to": "0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D",
            "value": "0x1BC16D674EC80000",
            "data": "0xb6f9de95000000000000000000000000000000000000000000000000000000dc2702dfd000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000c46dadbc6271511f25273ce856ecf799022da630000000000000000000000000000000000000000000000000000000064b818c30000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc20000000000000000000000005c72bb4774d70058b7761cc963782aa2826c7b81",
            "gas": "0x47CED",
            "maxFeePerGas": "0xF9D3F8FC3B",
            "maxPriorityFeePerGas": "0xE8990A4600",
        },
        "0x10e8465",
        {
            "timeout": "120s",
            "tracer": "callTracer",
        },
    ],
}


r = requests.post("http://geth:8545", json=req).json()

print(int(r["result"]["gasUsed"], 16))
```

**Comment by holiman** [comment]
The example transaction contains the `TIMESTAMP` opcode, which affects the contract flow. Any op that accesses the block context (timestamp, blocknumber etc), will be different when executed on _the block before_, and thus the gas used will be different. 
This is not an error. IF you want to trace the tx, you need to trace it differently, e.g. use `traceTransaction`.

**Comment by karalabe** [comment]
There are two ways to trace stuff: traceCall and traceTransaction. The latter reruns the transaction in the exact same context it was run while executing it originally. The traceCall is slightly different, because you can provide it a starting context on top of which to run, but it's kind of hard to manually make sure that everything is **exactly** the same (you need the parent state, you need the current block timestamp and number, etc). IT would probably even be impossible to do for transactions that are not the first in a block, because the pre-state for those would need rerunning everything before them.

TL;DR: If you want to trace transactions from blocks, you should use traceTransaction as that will make sure everything's set up correctly. If you want to trace non-existing transactions (or simply non-transaction VM executions), then use traceCall.

**Comment by geomad** [comment]
> > The example transaction contains the `TIMESTAMP` opcode, which affects the contract flow.
> 
> After providing the `blockOverrides`, the gasUsed is the same to traceTransaction
> 
> ```json
> {
>   "jsonrpc": "2.0",
>   "method": "debug_traceCall",
>   "id": 67,
>   "params": [
>     {
>       "from": "0x0C46dADBc6271511F25273cE856eCF799022dA63",
>       "to": "0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D",
>       "value": "0x1bc16d674ec80000",
>       "data": "0xb6f9de95000000000000000000000000000000000000000000000000000000dc2702dfd000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000c46dadbc6271511f25273ce856ecf799022da630000000000000000000000000000000000000000000000000000000064b818c30000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc20000000000000000000000005c72bb4774d70058b7761cc963782aa2826c7b81",
>       "gas": "0x47ced",
>       "gasPrice": "0xf3f0fd31a0"
>     },
>     "0x10e8465",
>     {
>       "timeout": "120s",
>       "tracer": "callTracer",
>       "blockOverrides": {
>         "number": "0x10e8466",
>         "time": "0x64b81857"
>       }
>     }
>   ]
> }
> ```

This worked perfectly!
I didn't know number and time could be overridden, since I didn't find `blockOverrides` anywhere mentioned in the docs. @karalabe maybe it should be added?
